10_rest/willshen8/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/anz-bank/go-course/10_rest/willshen8/pkg/puppy"
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var out io.Writer = os.Stdout
+
+var (
+	args     = os.Args[1:]
+	fileName = kingpin.Flag("data", "Puppy Json File Name").Short('d').File()
+	port     = kingpin.Flag("port", "Port of the server").Short('p').Default("8888").Int()
+	storer   = kingpin.Flag("store", "Map/Sync Store").Short('s').Default("map").Enum("map", "sync")
+)
+
+// Config contains arguments parsed from commandline flags.
+type Config struct {
+	file   *os.File
+	port   int
+	storer string
+}
+
+func main() {
+	cmdConfigs := parseCmdArgs(args)
+	newStorer, storerErr := createStorer(cmdConfigs)
+	if storerErr != nil {
+		panic(storerErr)
+	}
+	decodedPuppies, parseErr := parseJSONPuppies(cmdConfigs.file)
+	if parseErr != nil {
+		panic(parseErr)
+	}
+	portFlag, portErr := parsePortFlag(cmdConfigs)
+	if portErr != nil {
+		fmt.Println(portErr)
+	}
+	port := ":" + portFlag
+
+	puppyStorer, _ := createPuppies(newStorer, decodedPuppies)
+	fmt.Fprintln(out, puppyStorer)
+	puppyHandler := puppy.NewRestHandler(puppyStorer)
+	router := SetupRouter()
+	puppy.SetupRoutes(router, *puppyHandler)
+	log.Fatal(http.ListenAndServe(port, router))
+}
+
+// creatStorer will take the commandline arg and create the appropriate store
+func createStorer(c *Config) (puppy.Storer, error) {
+	switch c.storer {
+	case "map":
+		return puppy.NewMapStore(), nil
+	case "sync":
+		return puppy.NewSyncStore(), nil
+	}
+	return nil, errors.New("map/sync are the only acceptable flag values")
+}
+
+func parseCmdArgs(args []string) *Config {
+	if _, parseError := kingpin.CommandLine.Parse(args); parseError != nil {
+		panic(parseError)
+	}
+	return &Config{file: *fileName, port: *port, storer: *storer}
+}
+
+func parseJSONPuppies(r io.Reader) ([]puppy.Puppy, error) {
+	var puppies []puppy.Puppy
+	jsonData, readErr := ioutil.ReadAll(r)
+	if readErr != nil {
+		return nil, readErr
+	}
+
+	if err := json.Unmarshal(jsonData, &puppies); err != nil {
+		return nil, err
+	}
+	return puppies, nil
+}
+
+// parsePortFlag check for valid port number entered from commandLine
+func parsePortFlag(c *Config) (string, error) {
+	if c.port < 0 || c.port > 65535 {
+		return "0", errors.New("invalid port number entered")
+	}
+	return strconv.Itoa(c.port), nil
+}
+
+// createPuppies takes an array of puppies and saves it to a puppy storer
+func createPuppies(s puppy.Storer, decodedPuppies []puppy.Puppy) (puppy.Storer, error) {
+	for _, jsonPuppy := range decodedPuppies {
+		jsonPuppy := jsonPuppy //fixes linter issue
```

Comment:
nit: in some cases your using `blah //Comment` in other `blah // Comment` (space after slash). Please try to unify this.
Hint: `grep -r "//" .`

Reply:
Thanks, learning bash is on my list of to do. Appreciate that.

Reply:
Try and find a way to loop the following over every file
```bash
awk -F "//" -v OFS='' '{ if (substr($2, 1, 1) ~ /^[A-Za-z]/) $2="// "$2; print $0 }' file >tmp && mv tmp file
```

Reply:
a good editor should be able to look after that too.

Reply:
What editor do you use? I have used both VS code and Goland, or is it a setting issue?